### ISSUE TITLE
OCPI 2.2 including VAT makes it Europe-specific

### ISSUE THREAD

**Issue by KevBelisle** [ocpi/ocpi#211]
I have noticed that in the OCPI 2.2 specification (not sure how close it is to final) there are many places that reference VAT - notably, the [PriceComponent class](https://github.com/ocpi/ocpi/blob/develop/mod_tariffs.asciidoc#142-pricecomponent-class).

I think including taxes in the OCPI standard is a good idea, however the current implementation might be insufficient for certain countries.

For example, in Canada and the United States there are often multiple sales taxes to be applied.

A possible solution might be to use a list of 0-n instances of a `Tax` class instead of a simple `number` for the `PriceComponent.vat` property.

A `Tax` class might look like this :

- percentage, number, applicable tax percentage
- description, DisplayText, description of this tax

**Comment by simonschllng** [comment]
Is it always the total sum a tax needs to be applied to? Or are there countries/scenarios where even your proposal does not fit?

**Comment by KevBelisle** [comment]
In all cases I’ve seen so far, the same taxes are applied to the whole transaction.

I imagine it is possible that different taxes could be applied to different CdrDimensions.

**Comment by RobertdeLeeuw-IHomer** [comment]
I'm not an expert on VAT. When I was adding VAT to OCPI, I talked to a friend that works for the Dutch Tax Office (so this might only apply to Dutch law). She made it clear to me that we should not add too much details about the VAT in the CDR. The CDR is not an official invoice, the CPO should send an official invoice (not via OCPI) to the eMSP for one or more charging sessions. That invoice should contain all the VAT details like percentages, with possibly multiple different taxes. Why not in OCPI, then it could be seen as an invoice and be used for VAT deduction, and when the real invoice is then send a company could double deduct the VAT. Why do we want VAT in OCPI? EV drivers need to be shown costs including VAT, by sending both the price including VAT and excluding VAT, that should be possible for the eMSP.

**Comment by KevBelisle** [comment]
In Canada, there are certain requirements to show the end user which taxes he has paid, including the business’ registered tax numbers.

To meet that requirement, we “extended” OCPI 2.1 to add a Tax object to CDRs with that information (a list a tax objects, each comprised of a DisplayText description and an amount).




I think what I’m understanding is that in Europe you work things out between CPOs and eMSPs a little differently.

**Comment by RobertdeLeeuw-IHomer** [comment]
Is the CDR used as invoice from CPO to eMSP? then this is needed. If there is another invoice send, we have to think what is the best way to show enough information to the customer. Do have have somebody working in your company that has a lot of knowledge about VAT laws in Canada?

**Comment by RobertdeLeeuw-IHomer** [comment]
@KevBelisle any updates on this? Can I close this issue?

**Comment by KevBelisle** [comment]
Yes, we do have an accounting department which should be able to answer most questions regarding VAT laws in Canada/USA. We consulted with them when we implemented OCPI, I’ll try and get any information that would help OCPI be more “global-ready”.

I can also share the “taxes” property that we added to CDRs that describes any taxes included in the total_cost.

**Comment by RobertdeLeeuw-IHomer** [comment]
When you talk to them, explain that the CDR is not the B2B invoice, that is always send via some other way like e-mail.

**Comment by ali-gs** [comment]
Does OCPI support option if the TAX is calculated on Total or unit? if not can we add it in 2.2?
 
e.g. 0.42 per kWh Tax @20%
Option 1. Calculate tax on individual unit price
This will add the TAX rate to the individual unit price of the item, and then multiply the gross by the quantity.
Option 2. Calculate tax on total
This will multiply the quantity by the rate before applying the tax. 


kWh | Price | Tax 1 | Cost 1 | Tax 2 | Cost 2
-- | -- | -- | -- | -- | --
1 | 0.42 | 0.08 | 0.5 | 0.08 | 0.5
5 | 2.1 | 0.4 | 2.5 | 0.42 | 2.52

**Comment by Namoshek** [comment]
You can give a VAT value for each `PriceComponent` (page 113 in OCPI 2.2-RC2). A `PriceComponent` describes one `Dimension` of a `TariffElement`, e.g. `ENERGY` or `FLAT`. That means you can have multiple tax values per tariff.

An example of a Tariff from the documentation:
```
{
  "country_code": "DE",
  "party_id": "ALL",
  "id": "21",
  "currency": "EUR",
  "elements": [{
    "price_components": [{
      "type": "TIME",
      "price": 3.00,
      "vat": 10.0,
      "step_size": 60
    }, {
      "type": "PARKING_TIME",
      "price": 5.00,
      "vat": 20.0,
      "step_size": 300
    }]
  }],
  "last_updated": "2018-12-17T17:00:43Z"
}
```
In this example, parking time will be taxed with 20% and charging time with 10%.

**Comment by ali-gs** [comment]
> You can give a VAT value for each `PriceComponent` (page 113 in OCPI 2.2-RC2). A `PriceComponent` describes one `Dimension` of a `TariffElement`, e.g. `ENERGY` or `FLAT`. That means you can have multiple tax values per tariff.
> 
> An example of a Tariff from the documentation:
> 
> ```
> {
>   "country_code": "DE",
>   "party_id": "ALL",
>   "id": "21",
>   "currency": "EUR",
>   "elements": [{
>     "price_components": [{
>       "type": "TIME",
>       "price": 3.00,
>       "vat": 10.0,
>       "step_size": 60
>     }, {
>       "type": "PARKING_TIME",
>       "price": 5.00,
>       "vat": 20.0,
>       "step_size": 300
>     }]
>   }],
>   "last_updated": "2018-12-17T17:00:43Z"
> }
> ```
> 
> In this example, parking time will be taxed with 20% and charging time with 10%.

Yes I understand the PriceComponent part but my question is in method of calculation of vat on the same price component.

**Comment by Namoshek** [comment]
Does it matter? I think it only matters if you round between each step in the calculation. Multiplication itself obviously doesn't care about the order of factors:
```
5 x 0.42 x 1.08   =   5 x 1.08 x 0.42   =   0.42 x 5 x 1.08   =   0.42 x 1.08 x 5
```
whereas
```
round(5 x 0.42, 2) x 1.08   !=   round(0.42 x 1.08, 2) x 5
```

So my answer would be that you should only round when you need to store the calculated costs, not during the calculation itself. But I'm curious what others think about this.

**Comment by thanaParis** [comment]
Continuing this discussion, as something similar has come up in my use case.

We use CDRs to inform the customer about the cost of the session. There is a separate B2B invoice which is used to settle payment, but it is not displayed to the end user. Only the CDR is.

I think a tax breakdown in the CDR would be very useful for this purpose alone. Demanding that CDRs not be used for B2B settlement doesn't preclude the benefit of showing EV drivers the actual cost of their session using OCPI.

This is separate from including tax information in the Tariff. Tariff tax rates are useful to show expected costs, they are not a replacement for showing the user their final cost including taxes.
